xls/modules/rle/rle.x

```diff
@@ -0,0 +1,253 @@
+// Copyright 2023 The XLS Authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+// This file implements a parametric RLE encoder
+//
+// The encoder uses Run Length Encoding (RLE) to compress the input stream
+// with symbols (data to compress). The output contains packets of data with
+// the symbols and the number of their occurrences in the input stream.
+
+import std
+
+// structure containing output data from an RLE encoder
+struct EncOutData<SYMB_WIDTH: u32, CNT_WIDTH: u32> {
+    symb: bits[SYMB_WIDTH], // symbol
+    cnt: bits[CNT_WIDTH],   // symbol counter
+}
+
+// structure to preserve the state of an RLE encoder for the next iteration
+pub struct RLEEncState<SYMB_WIDTH: u32, CNT_WIDTH: u32> {
+    prev: EncOutData<SYMB_WIDTH, CNT_WIDTH>, // previous symbol and symbol counter
+    flush: bool,                             // if flush was requested
```

Comment:
I'm curious about the semantic of the `flush` and why it needs to be kept as a state, could you add some rational in the PR description or in the code comment?

I would have naively thought that the encoder would emit on every symbol change, but it seems that you are giving more control with that flag.

Reply:
A request on the `flush` channel forces the encoder to dump all data to the output stream. Both the `flush` and the channel with input symbols should be synchronized. The information about the flush request is stored in the state of the encoder because it is used to skip reading symbols in the next cycle after the request. That "skipped" cycle is used to send old data (about the last symbol when flush occurred). 

This implementation was chosen because it is currently impossible to send data two times in a row to the same channel in one evaluation of the `next` function.

The typical use of this channel is to set flush after all symbols have been sent to force the encoder to output a final result. Otherwise, the last symbol will be "stuck" in the encoder.

Of course, I will update the code with the proper explanation about the `flush` channel and the `flush` in the state.

Reply:
> This implementation was chosen because it is currently impossible to send data two times in a row to the same channel in one evaluation of the next function.

Are you referring to https://github.com/google/xls/blob/30f192cb091f3f36e459bf4738b4447543f9177a/xls/ir/verifier.cc#L1604 ?

Curious if https://github.com/google/xls/blob/main/docs_src/design_docs/legalize_multiple_channel_ops_per_channel.md would help /cc @grebe 

Interestingly it seems that it doesn't currently fails the interpreter, nor IR conversion, but IR optimization, see:
https://colab.research.google.com/gist/proppy/9be8ce46181b007d1cbe2a4329923623/dslx-proc-playground.ipynb

Reply:
Yes, that is the problem that I encountered.